src/server/model_outputs.R

```diff
@@ -8,22 +8,84 @@ fitModel <- function(maxIterations, likelihood, spectrumData) {
         maxIterations <- 2
     }
 
-    optim(
-        theta0,
-        first90::ll_hts,
-        fp = spectrumData,
-        likdat = likelihood,
-        method = "BFGS",
-        control = list(fnscale = -1, trace=4, REPORT=1, maxit=maxIterations),
-        hessian = !testMode
-    )
+    opt <- NULL
+
+    shiny::withProgress(message = 'Fitting model', value = 0, {
+
+        i <- Counter$new(par = 0, iteration = 0)
+
+        opt <- optim(
+            theta0,
+            iterate,
+            fp = spectrumData,
+            likdat = likelihood,
+            maxRuns = maxIterations*2*length(theta0), # this is a heuristic for how many times the model is run by optim
+            i = i,
+            method = "BFGS",
+            control = list(fnscale = -1, trace=4, REPORT=1, maxit=maxIterations)
+        )
+
+    })
+
+    opt
 }
 
-runSimulations <- function(opt, spectrumData) {
-    testMode <- Sys.getenv("SHINY90_TEST_MODE") == "TRUE"
-    if (testMode) {
-        NULL
-    } else {
-        first90::simul.test(opt, spectrumData)
+iterate <- function(theta, fp, likdat, maxRuns, i){
+
+    val <- first90::ll_hts(theta, fp, likdat)
+
+    if (isGradientStep(theta, i$par) || is.nan(val)){
+        shiny::incProgress((1/(maxRuns)))
+    }
+    else {
+        shiny::incProgress((1/(maxRuns)), detail= paste("Converging: ", round(val, 2)))
+    }
+
+    i$par <- theta
+    i$iteration <- i$iteration + 1
+    val
+}
+
+isGradientStep <- function(theta1, theta2){
+    length(which(theta1 != theta2)) < 2
+}
+
+iterateHessian <- function(theta, fp, likdat, i){
+
+    shiny::incProgress(1/7300) # 7300 is an approximation of how many times the model will be run to calculate the hessian
+    i$iteration <- i$iteration + 1
+
+    first90::ll_hts(theta, fp, likdat)
+}
+
+make_progress <- function(n) {
+    counter <- 0L
+    function() {
+        counter <<- counter + 1L
+        shiny::incProgress(1/n, detail = glue::glue("Running simulation {counter} of {n}"))
```

Comment:
Something going on here I can't work out - this message gets up to "Running simulation n+1 of n" - but I can't see why?!

Reply:
We worked out that this was an `rbind` of the parameters onto the number of simulations within the first90 function